sway/tree/container.c

```diff
@@ -596,14 +597,43 @@ struct sway_container *container_at(struct sway_container *parent,
 	case L_STACKED:
 		return container_at_stacked(parent, ox, oy, surface, sx, sy);
 	case L_FLOATING:
-		return NULL; // TODO
+		sway_assert(false, "Didn't expect to see floating here");
+		return NULL;
 	case L_NONE:
 		return NULL;
 	}
 
 	return NULL;
 }
 
+struct sway_container *floating_container_at(double lx, double ly,
+		struct wlr_surface **surface, double *sx, double *sy) {
+	for (int i = 0; i < root_container.children->length; ++i) {
+		struct sway_container *output = root_container.children->items[i];
+		for (int j = 0; j < output->children->length; ++j) {
+			struct sway_container *workspace = output->children->items[j];
+			struct sway_workspace *ws = workspace->sway_workspace;
+			if (!workspace_is_visible(workspace)) {
+				continue;
+			}
+			for (int k = 0; k < ws->floating->children->length; ++k) {
+				struct sway_container *floater =
+					ws->floating->children->items[k];
+				struct wlr_box box = {
+					.x = floater->x,
+					.y = floater->y,
+					.width = floater->width,
+					.height = floater->height,
+				};
+				if (wlr_box_contains_point(&box, lx, ly)) {
```

Comment:
I floated Firefox, then right clicked near the right side so the context menu popped up outside of the float. I can interact with the popup outside the floating area just fine. That's because layer surfaces are checked before `floating_container_at` and `container_at` are called.

Reply:
This is completely unrelated to layer surfaces. Firefox is xwayland, which uses unmanaged surfaces that are global to the whole layout and don't have a parent. I'm talking about xdg-popups and wayland subsurfaces.

Reply:
I can replicate this with FileZilla. However, I can also replicate it when tiled, and I can replicate in master, and I can replicate it in master from a week ago which was prior to when I split `container_at` into separate functions. So this problem has been around for a while.

I guess the solution will be to iterate the containers, check their subsurfaces first, then iterate them all again and check the toplevel surface. The same has to happen for rendering too - currently a tiled view can render over the top of its left sibling's context menu.

My preference is to make a separate PR for this because it's not specific to floating.

Reply:
Filezilla is *not* using xdg-shell. Pick any GTK+3 app, right click, and you should get a popup.

Reply:
Bump

Reply:
Hmm, can you just remove this `wlr_box_contains_point` check and only use `container_at`?